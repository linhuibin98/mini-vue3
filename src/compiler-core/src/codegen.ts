import { isString } from '../../shared'
import { NodeTypes } from './ast'
import { CREATE_ELEMENT_VNODE, TO_DISPLAY_STRING, helperMapName } from './runtimeHelper'

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context

  // import
  if (ast.helpers.length > 0) {
    const vueBinging = 'Vue'
    const aliasHelper = v => `${helperMapName[v]}: _${helperMapName[v]}}`

    push(`const { ${ast.helpers.map(aliasHelper).join(', ')} } = ${vueBinging}`)
    push('\n')
  }

  // return
  push('return ')

  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}){`)
  push('return ')

  // ast 生成 code
  genNode(ast.codegenNode, context)

  push('}')

  return {
    code: context.code,
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    },
  }

  return context
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
      break
    default:
      break
  }
}

function genCompoundExpression(node, context) {
  const { children } = node
  const { push } = context
  for (const childNode of children) {
    if (isString(childNode))
      push(childNode)

    else
      genNode(childNode, context)
  }
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, children, props } = node
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  genNodeList(genNullable([tag, props, children]), context)
  push(')')
}

function genNodeList(nodes, context) {
  const { push } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isString(node))
      push(node)

    else
      genNode(node, context)

    if (i < nodes.length - 1)
      push(', ')
  }
}

function genNullable(args) {
  return args.map(arg => arg || 'null')
}

function genText(node, context) {
  const { push } = context

  push(`'${node.content}'`)
}

function genInterpolation(node, context) {
  const { push, helper } = context

  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

function genExpression(node, context) {
  const { push } = context

  push(node.content)
}
