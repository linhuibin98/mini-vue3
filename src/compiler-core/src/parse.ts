import { NodeTypes } from './ast'

interface Context {
  source: string
}

const enum TagType {
  START,
  END,
}

export function baseParse(content: string) {
  const context = createParserContext(content)
  return createRoot(parseChildren(context, []))
}

function parseChildren(context: Context, ancestors: string[]) {
  const nodes: any[] = []

  while (!isEnd(context, ancestors)) {
    let node
    const s = context.source
    if (s.startsWith('{{')) {
      node = parseInterpolation(context)
    }
    else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1]))
        node = parseElement(context, ancestors)
    }

    if (!node)
      node = parseText(context)

    nodes.push(node)
  }

  return nodes
}

function isEnd(context: Context, ancestors: string[]) {
  const s = context.source
  // 是否是结束标签
  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      if (startsWidthEndTag(s, ancestors[i]))
        return true
    }
  }
  return s.length === 0
}

function parseText(context: Context) {
  const s = context.source
  const l = s.length
  let endIndex = l
  const endTokens = ['{{', '</']

  for (const t of endTokens) {
    const index = s.indexOf(t)

    if (index !== -1 && index < endIndex)
      endIndex = index
  }

  const content = parserTextData(context, endIndex)
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseElement(context: Context, ancestors: string[]) {
  // 处理开始标签
  const element = parseTag(context, TagType.START)!
  // 保存标签解析栈
  ancestors.push(element.tag)

  // 孩子节点递归
  element.children = parseChildren(context, ancestors)
  // 标签处理结束，移出栈
  ancestors.pop()

  // 处理结束标签
  if (startsWidthEndTag(context.source, element.tag))
    parseTag(context, TagType.END)

  else
    throw new Error(`缺少结束标签：${element.tag}`)

  return element
}

function startsWidthEndTag(source: string, tag: string) {
  return source.startsWith('</') && tag === source.slice(2, 2 + tag.length)
}

function parseTag(context, type) {
  const s = context.source
  // 1. 解析 tag
  const match = /^<\/?([a-z]*)/i.exec(s)
  if (!match) return
  const tag = match[1]
  // 2. 删除处理完成代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)
  if (type === TagType.END) return

  return {
    type: NodeTypes.ELEMENT,
    tag,
    children: [] as any,
  }
}

function parseInterpolation(context) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(closeDelimiter)

  advanceBy(context, openDelimiter.length)

  const rawContentLength = closeIndex - openDelimiter.length

  const content = parserTextData(context, rawContentLength)

  advanceBy(context, closeDelimiter.length)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: content.trim(),
    },
  }
}

function parserTextData(context, length) {
  const content = context.source.slice(0, length)
  advanceBy(context, length)
  return content
}

function advanceBy(context: Context, length: number) {
  context.source = context.source.slice(length)
}

function createRoot(children) {
  return {
    children,
  }
}

function createParserContext(content: string) {
  return {
    source: content,
  }
}
