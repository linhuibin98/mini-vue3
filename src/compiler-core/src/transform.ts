import { NodeTypes } from './ast'
import { CREATE_ELEMENT_VNODE, TO_DISPLAY_STRING } from './runtimeHelper'

export function transform(root, options = {}) {
  const context = createTransformContext(root, options)
  // 1. 遍历 - 深度优先
  traverseNode(root, context)

  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
  const childNode = root.children[0]
  if (childNode.type === NodeTypes.ELEMENT)
    root.codegenNode = childNode.codegenNode

  else
    root.codegenNode = root.children[0]
}

function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms
  const exitFns: any[] = []

  for (const transformFn of nodeTransforms) {
    const exitFn = transformFn(node, context)
    if (exitFn) exitFns.push(exitFn)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node.children, context)
      break
    case NodeTypes.TEXT:
      context.helper(CREATE_ELEMENT_VNODE)
      break
    default:
      break
  }

  let i = exitFns.length

  while (i--)
    exitFns[i]()
}

function traverseChildren(children, context) {
  if (!children) return

  for (const childNode of children)
    traverseNode(childNode, context)
}

function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      if (!context.helpers.has(key))
        context.helpers.set(key, 1)
    },
  }

  return context
}
