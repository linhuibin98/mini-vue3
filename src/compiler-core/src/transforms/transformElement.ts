import { NodeTypes, createVNodeCall } from '../ast'

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      // 中间层处理

      // tag
      const vnodeTag = `'${node.tag}'`
      // props
      let vnodeProps
      // children
      const children = node.children
      const vnodeChildren = children[0]

      node.codegenNode = createVNodeCall(context, vnodeTag, vnodeProps, vnodeChildren)
    }
  }
}

