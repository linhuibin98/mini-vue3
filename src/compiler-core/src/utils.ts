import { NodeTypes } from './ast'

export const isText = node => [NodeTypes.TEXT, NodeTypes.INTERPOLATION].includes(node.type)
