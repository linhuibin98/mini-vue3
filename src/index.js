import { reactive } from './reactivity';

export default {
    reactive
};