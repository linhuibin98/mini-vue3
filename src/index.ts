export * from './reactivity'
export * from './shared'
export * from './runtime-core'
export * from './runtime-dom'
