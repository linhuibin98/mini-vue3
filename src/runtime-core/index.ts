export * from './createApp'
export * from './vnode'
export * from './renderer'
export * from './helpers/renderSlots'
