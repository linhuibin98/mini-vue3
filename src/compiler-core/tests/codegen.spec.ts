import { baseParse } from '../src'
import { generate } from '../src/codegen'
import { transform } from '../src/transform'
import { transformExpression } from '../src/transforms/transformExpression'

describe('codegen', () => {
  it('gen text', () => {
    const source = 'Hi'
    const ast = baseParse(source)

    transform(ast)

    expect(generate(ast)).toMatchSnapshot()
  })

  it('gen interpolation', () => {
    const source = '{{message}}'
    const ast = baseParse(source)

    transform(ast, {
      nodeTransforms: [transformExpression],
    })

    expect(generate(ast)).toMatchSnapshot()
  })
})
