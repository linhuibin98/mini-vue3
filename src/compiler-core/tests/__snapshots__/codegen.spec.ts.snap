// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen gen element 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode} } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, null)}",
}
`;

exports[`codegen gen element and text and interpolation 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode}, toDisplayString: _toDisplayString} } = Vue
return function render(_ctx, _cache){return _createElementVNode('div', null, 'Hi,' + _toDisplayString(_ctx.message))}",
}
`;

exports[`codegen gen interpolation 1`] = `
Object {
  "code": "const { toDisplayString: _toDisplayString} } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}",
}
`;

exports[`codegen gen text 1`] = `
Object {
  "code": "const { createElementVNode: _createElementVNode} } = Vue
return function render(_ctx, _cache){return 'Hi'}",
}
`;
